art/attacks/evasion/pixel_threshold.py

```diff
@@ -0,0 +1,1346 @@
+# MIT License
+#
+# Copyright (C) IBM Corporation 2018
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements the Threshold Attack and Pixel Attack.
+The Pixel Attack is a generalisation of One Pixel Attack.
+
+| One Pixel Attack Paper link:
+    https://ieeexplore.ieee.org/abstract/document/8601309/citations#citations
+    (arXiv link: https://arxiv.org/pdf/1710.08864.pdf)
+| Pixel and Threshold Attack Paper link:
+    https://arxiv.org/abs/1906.06026
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+# import warnings
+import logging
+from itertools import product
+
+import numpy as np
+
+from scipy._lib.six import xrange, string_types
+from scipy._lib._util import check_random_state
+from scipy.optimize.optimize import _status_message
+from scipy.optimize import OptimizeResult, minimize
+
+from cma import CMAEvolutionStrategy, CMAOptions
+# from scipy.optimize import differential_evolution
```

Comment:
```suggestion
```

Reply:
Kindly let it be here. The current scipy's version of differential evolution doesn't work but maybe in future I or someone else can remove the code included below the method and use this line. 
This line will act as a reminder to update the code.

Reply:
Actually the code uses a modified script of SciPy's implementation of Differential Evolution Optimisation. However, the best case would be to make the DE optimisation compatible with the current SciPy's implementation.
At the current stage, I have not yet identified all the changes in the implementation but it can be done at a later stage.

_Please let it be here so that later I or someone else can update this. This comment ideally would hint that a version of SciPy's Differential Evolution is used._

Reply:
Ok, could you please add this explanation as comment after line 45?

Reply:
I have added an explanation around the import. Please let me know if it works or needs to be modified.

Reply:
Looks good.